File: backend/router/account.py
from fastapi import APIRouter, Request, Response, Cookie, status, Depends
from fastapi.responses import JSONResponse
from pydantic import BaseModel, EmailStr
from typing import Optional
from env import DB_PATH
import sqlite3

router = APIRouter()

def get_db():
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

def try_catch(fn):
    try:
        return {"data": fn(), "error": None}
    except Exception as e:
        return {"data": None, "error": str(e)}

def auth(token: Optional[str] = Cookie(None)):
    if not token:
        return None
    conn = get_db()
    cur = conn.cursor()
    cur.execute("SELECT * FROM sessions WHERE token = ?", (token,))
    session = cur.fetchone()
    if not session:
        return None
    cur.execute("SELECT * FROM users WHERE id = ?", (session["user_id"],))
    user = cur.fetchone()
    conn.close()
    if not user:
        return None
    return {"session": session, "user": user}

# Request Models
class DeleteAccountRequest(BaseModel):
    password: str

class ChangePasswordRequest(BaseModel):
    old_password: str
    new_password: str

class ChangeEmailRequest(BaseModel):
    email: EmailStr
    password: str

class ChangeNameRequest(BaseModel):
    full_name: str

@router.delete("/api/account")
def delete_account(body: DeleteAccountRequest, token: Optional[str] = Cookie(None)):
    print("INFO: DELETE /api/account called")
    session = auth(token)
    if not session:
        print("ERROR: Authentication failed: Unauthorized")
        return JSONResponse(status_code=401, content={"error": "Unauthorized"})

    user = session["user"]
    if user["password"] != body.password:
        print("WARN: Account deletion failed: Incorrect password")
        return JSONResponse(status_code=401, content={"error": "Incorrect password"})

    conn = get_db()
    cur = conn.cursor()
    result = try_catch(lambda: cur.execute("DELETE FROM users WHERE id = ?", (user["id"],)))
    if result["error"]:
        print(f"ERROR: {result['error']}")
        return JSONResponse(status_code=500, content={"error": "Internal server error"})

    conn.commit()
    conn.close()
    print(f"INFO: User deleted: {user['id']}")
    response = JSONResponse(content={"message": "Account deleted"})
    response.delete_cookie("token")
    return response

@router.patch("/api/account/chpwd")
def change_password(body: ChangePasswordRequest, token: Optional[str] = Cookie(None)):
    print("INFO: PATCH /api/account/chpwd called")
    session = auth(token)
    if not session:
        print("ERROR: Authentication failed: Unauthorized")
        return JSONResponse(status_code=401, content={"error": "Unauthorized"})

    user = session["user"]
    if user["password"] != body.old_password:
        print("WARN: Password change failed: Incorrect old password")
        return JSONResponse(status_code=401, content={"error": "Incorrect old password"})

    conn = get_db()
    cur = conn.cursor()
    result = try_catch(lambda: cur.execute(
        "UPDATE users SET password = ? WHERE id = ?", (body.new_password, user["id"])
    ))
    if result["error"]:
        print(f"ERROR: {result['error']}")
        return JSONResponse(status_code=500, content={"error": "Internal server error"})

    conn.commit()
    conn.close()
    print(f"INFO: Password changed for user: {user['id']}")
    return {"message": "Password changed"}

@router.patch("/api/account/chemail")
def change_email(body: ChangeEmailRequest, token: Optional[str] = Cookie(None)):
    print("INFO: PATCH /api/account/chemail called")
    session = auth(token)
    if not session:
        print("ERROR: Authentication failed: Unauthorized")
        return JSONResponse(status_code=401, content={"error": "Unauthorized"})

    user = session["user"]
    if user["password"] != body.password:
        print("WARN: Email change failed: Incorrect password")
        return JSONResponse(status_code=401, content={"error": "Incorrect password"})

    conn = get_db()
    cur = conn.cursor()
    result = try_catch(lambda: cur.execute(
        "UPDATE users SET email = ? WHERE id = ?", (body.email, user["id"])
    ))
    if result["error"]:
        print(f"ERROR: {result['error']}")
        return JSONResponse(status_code=500, content={"error": "Internal server error"})

    conn.commit()
    conn.close()
    print(f"INFO: Email changed for user: {user['id']}")
    return {"message": "Email changed"}

@router.patch("/api/account/chname")
def change_name(body: ChangeNameRequest, token: Optional[str] = Cookie(None)):
    print("INFO: PATCH /api/account/chname called")
    session = auth(token)
    if not session:
        print("ERROR: Authentication failed: Unauthorized")
        return JSONResponse(status_code=401, content={"error": "Unauthorized"})

    conn = get_db()
    cur = conn.cursor()
    result = try_catch(lambda: cur.execute(
        "UPDATE users SET full_name = ? WHERE id = ?", (body.full_name, session["user"]["id"])
    ))
    if result["error"]:
        print(f"ERROR: {result['error']}")
        return JSONResponse(status_code=500, content={"error": "Internal server error"})

    conn.commit()
    conn.close()
    print(f"INFO: Name changed for user: {session['user']['id']}")
    return {"message": "Name changed"}

File: express/src/router/account.ts
import { Router } from "express";
import { auth } from "../utils/auth";
import { tryCatch } from "../utils/trycatch";
import { db } from "../db";
import z from "zod";
import getLogger from "../utils/logger";
import { users } from "../db/schema";
import { eq } from "drizzle-orm";

export function accountRouter() {
    const router = Router();
    const logger = getLogger("router.account");

    router.delete("/api/account", async (req, res) => {
        logger("INFO", "DELETE /api/account called");

        const authRes = await auth(req);

        if (!authRes) {
            logger("ERROR", "Authentication failed: Unauthorized");
            res.status(401).json({ error: "Unauthorized" });
            return;
        }

        const bodySchema = z.object({
            password: z.string(),
        });
        const body = bodySchema.safeParse(req.body);
        if (!body.success) {
            logger("WARN", `Validation failed: ${JSON.stringify(body.error.errors)}`);
            res.status(400).json({ error: body.error.errors });
            return;
        }
        const { password } = body.data;
        const userId = authRes.user.id;
        const user = await tryCatch(db.select().from(users).where(eq(users.id, userId)));
        if (user.error) {
            logger("ERROR", `DB error fetching user: ${user.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        if (user.data.length === 0) {
            logger("ERROR", "User not found");
            res.status(404).json({ error: "User not found" });
            return;
        }
        const existingUser = user.data[0];
        if (existingUser.password !== password) {
            logger("WARN", "Account deletion failed: Incorrect password");
            res.status(401).json({ error: "Incorrect password" });
            return;
        }
        const result = await tryCatch(db.delete(users).where(eq(users.id, userId)));
        if (result.error) {
            logger("ERROR", `DB error deleting user: ${result.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        logger("INFO", `User deleted: ${userId}`);
        res.status(200).json({ message: "Account deleted" });
    });

    router.patch("/api/account/chpwd", async (req, res) => {
        logger("INFO", "PATCH /api/account/chpwd called");
        const authRes = await auth(req);
        if (!authRes) {
            logger("ERROR", "Authentication failed: Unauthorized");
            res.status(401).json({ error: "Unauthorized" });
            return;
        }
        const userId = authRes.user.id;
        const bodySchema = z.object({
            old_password: z.string(),
            new_password: z.string(),
        });
        const body = bodySchema.safeParse(req.body);
        if (!body.success) {
            logger("WARN", `Validation failed: ${JSON.stringify(body.error.errors)}`);
            res.status(400).json({ error: body.error.errors });
            return;
        }
        const { old_password, new_password } = body.data;
        const user = await tryCatch(db.select().from(users).where(eq(users.id, userId)));
        if (user.error) {
            logger("ERROR", `DB error fetching user: ${user.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        if (user.data.length === 0) {
            logger("ERROR", "User not found");
            res.status(404).json({ error: "User not found" });
            return;
        }
        const existingUser = user.data[0];
        if (existingUser.password !== old_password) {
            logger("WARN", "Password change failed: Incorrect old password");
            res.status(401).json({ error: "Incorrect old password" });
            return;
        }
        const result = await tryCatch(db.update(users).set({ password: new_password }).where(eq(users.id, userId)));
        if (result.error) {
            logger("ERROR", `DB error updating password: ${result.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        logger("INFO", `Password changed for user: ${userId}`);
        res.status(200).json({ message: "Password changed" });
    });

    router.patch("/api/account/chemail", async (req, res) => {
        logger("INFO", "PATCH /api/account/chemail called");
        const authRes = await auth(req);
        if (!authRes) {
            logger("ERROR", "Authentication failed: Unauthorized");
            res.status(401).json({ error: "Unauthorized" });
            return;
        }
        const userId = authRes.user.id;
        const bodySchema = z.object({
            email: z.string().email(),
            password: z.string(),
        });
        const body = bodySchema.safeParse(req.body);
        if (!body.success) {
            logger("WARN", `Validation failed: ${JSON.stringify(body.error.errors)}`);
            res.status(400).json({ error: body.error.errors });
            return;
        }
        const { email, password } = body.data;
        const user = await tryCatch(db.select().from(users).where(eq(users.id, userId)));
        if (user.error) {
            logger("ERROR", `DB error fetching user: ${user.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        if (user.data.length === 0) {
            logger("ERROR", "User not found");
            res.status(404).json({ error: "User not found" });
            return;
        }
        const existingUser = user.data[0];
        if (existingUser.password !== password) {
            logger("WARN", "Email change failed: Incorrect password");
            res.status(401).json({ error: "Incorrect password" });
            return;
        }
        const result = await tryCatch(db.update(users).set({ email }).where(eq(users.id, userId)));
        if (result.error) {
            logger("ERROR", `DB error updating email: ${result.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        logger("INFO", `Email changed for user: ${userId}`);
        res.status(200).json({ message: "Email changed" });
    });

    router.patch("/api/account/chname", async (req, res) => {
        logger("INFO", "PATCH /api/account/chname called");
        const authRes = await auth(req);
        if (!authRes) {
            logger("ERROR", "Authentication failed: Unauthorized");
            res.status(401).json({ error: "Unauthorized" });
            return;
        }
        const userId = authRes.user.id;
        const bodySchema = z.object({
            full_name: z.string(),
        });
        const body = bodySchema.safeParse(req.body);
        if (!body.success) {
            logger("WARN", `Validation failed: ${JSON.stringify(body.error.errors)}`);
            res.status(400).json({ error: body.error.errors });
            return;
        }
        const { full_name } = body.data;
        const result = await tryCatch(db.update(users).set({ full_name }).where(eq(users.id, userId)));
        if (result.error) {
            logger("ERROR", `DB error updating name: ${result.error}`);
            res.status(500).json({ error: "Internal server error" });
            return;
        }
        logger("INFO", `Name changed for user: ${userId}`);
        res.status(200).json({ message: "Name changed" });
    });

    return router;
}
